### ISSUE TITLE
shared-library-without-dependency-information false positives

### ISSUE THREAD

**Issue by jengelh** [rpm-software-management/rpmlint#969]
Given this shared library:

```
$ cat x.c
int compute_bla(int x) { return x * x * x; }
$ gcc -c x.c -Wall -O2
$ gcc -shared -o libx.so x.o -Wl,--as-needed
$ ldd libx.so
        statically linked
```

rpmlint complains like so

```
gap-datastructures.aarch64: E: shared-library-without-dependency-information /usr/lib64/gap/pkg/datastructures-0.3.0/bin/aarch64-suse-linux-gnu-default64-kv8/datastructures.so                                                       
```

openSUSE uses -Wl,--as-needed, and I seem to remember some other distro doing it too. So it's not like the compilation step has willingly used the -nostdlib that rpmlint was after.

**Comment by marxin** [comment]
Confirmed, however, I'm curious how can we distinguish in between a library that was really statically linked against `glibc` and one that doesn't use any symbol from it?

**Comment by jengelh** [comment]
>distinguish in between a library that was really statically linked against glibc

You could look for libc symbols in the shared object. But that would depend on being able to create such a monster… and in general, it does not appear like you can:

* amd64 is picky about PIC. Trying to link libc.a into a shared object won't work because of the insufficient reloc type:

  ```
  $ gcc -fPIC -shared rand_wrap.c -Wl,--as-needed,-Bstatic,-lc,-Bdynamic 
  /usr/lib64/gcc/x86_64-suse-linux/12/../../../../x86_64-suse-linux/bin/ld: /usr/lib64/gcc/x86_64-suse-linux/12/../../../../lib64/libc.a(malloc.o): relocation R_X86_64_TPOFF32 against `tcache' can not be used when making a shared object; recompile with -fPIC
  /usr/lib64/gcc/x86_64-suse-linux/12/../../../../x86_64-suse-linux/bin/ld: failed to set dynamic section sizes: bad value
  collect2: error: ld returned 1 exit status
  ```

  openSUSE has no `libc_pic.a`, so that's that.

* On i586, where PIC requirements are nil, this is still difficult:

  ```
  $ /usr/lib64/gcc/x86_64-suse-linux/12/collect2 -plugin /usr/lib64/gcc/x86_64-suse-linux/12/liblto_plugin.so -plugin-opt=/usr/lib64/gcc/x86_64-suse-linux/12/lto-wrapper -plugin-opt=-fresolution=/tmp/ccyheiYs.res -plugin-opt=-pass-through=-lgcc -plugin-opt=-pass-through=-lgcc_s -plugin-opt=-pass-through=-lc -plugin-opt=-pass-through=-lgcc -plugin-opt=-pass-through=-lgcc_s --build-id --eh-frame-hdr -m elf_i386 -shared /usr/lib64/gcc/x86_64-suse-linux/12/../../../../lib/crti.o /usr/lib64/gcc/x86_64-suse-linux/12/32/crtbeginS.o -L/usr/lib64/gcc/x86_64-suse-linux/12/32 -L/usr/lib64/gcc/x86_64-suse-linux/12/../../../../x86_64-suse-linux/lib/../lib -L/usr/lib64/gcc/x86_64-suse-linux/12/../../../../lib -L/lib/../lib -L/usr/lib/../lib -L/usr/lib64/gcc/x86_64-suse-linux/12 -L/usr/lib64/gcc/x86_64-suse-linux/12/../../../../x86_64-suse-linux/lib -L/usr/lib64/gcc/x86_64-suse-linux/12/../../.. rand_wrap.o /lib/libc.a  /usr/lib64/gcc/x86_64-suse-linux/12/32/crtendS.o /usr/lib64/gcc/x86_64-suse-linux/12/../../../../lib/crtn.o
  /usr/bin/ld: /lib/libc.a(malloc.o): warning: relocation in read-only section `.text'
  /usr/bin/ld: /lib/libc.a(libc-start.o): in function `call_fini':
  /home/abuild/rpmbuild/BUILD/glibc-2.36/csu/../csu/libc-start.c:196: undefined reference to `__fini_array_end'
  /usr/bin/ld: /lib/libc.a(libc-start.o): relocation R_386_GOTOFF against undefined hidden symbol `__fini_array_end' can not be used when making a shared object
  /usr/bin/ld: final link failed: bad value
  collect2: error: ld returned 1 exit status
  ```

* `__fini_array_end` is generated by ld from `/usr/lib64/ldscripts/*` and AFAICS only for program executables, not shared objects.

**Comment by marxin** [comment]
Right, it seems the majority of the "statically linked" libraries (as claimed by `lld`) are really just not dependent on another library.
Do you want to come up with a pull request that would remove the error? Is it what you're basically suggesting?
